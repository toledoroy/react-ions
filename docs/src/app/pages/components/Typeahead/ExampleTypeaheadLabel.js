import React from 'react'
import Typeahead from 'react-ions/lib/components/Typeahead/Typeahead'
import options from './CountryList'

const ExampleTypeaheadLabel = () => (
  <Typeahead
    options={options}
    valueProp='countryCode'
    displayProp='countryName'
    placeholder='Start typing'
    label='Typeahead Label'
    width='220px'
  />
)

export default ExampleTypeaheadLabel
