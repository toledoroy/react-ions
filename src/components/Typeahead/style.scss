@import '../../styles/common/colors';
@import '../../styles/common/screen';
@import '../../styles/common/mixins';

.typeahead-component {
  position: relative;

  label {
    color: $navy-text;
    display: block;
    font-size: 16px;
    font-weight: 200;
    margin-bottom: 10px;
  }

  .typeahead-list {
    background-color: $white;
    border: 2px solid $neutral-1;
    border-radius: 0 0 3px 3px;
    border-top: 0;
    color: $primary-4;
    list-style-type: none;
    margin: 0;
    max-height: 300px;
    overflow: auto;
    padding: 0;
    position: absolute;
    width: calc(100% - 2px) !important;
    z-index: 1;

    li {
      border-bottom: 1px solid $neutral-1;
      cursor: pointer;
      padding: 10px;

      &:last-child {
        border-bottom: 0;
      }

      &:hover {
        background-color: $primary-1;
        color: $white;
      }
    }
  }

  .input-wrapper {
    position: relative;
  }
  
  @media #{$media-extra-small} {
  .typeahead-component-styled {
      width: 100% !important;
    }
  }
}

.loader {
  position: absolute;
  right: 18px;
  top: 18px;
}

.reset-button {
  background-color: $neutral-4;
  border-radius: 50%;
  cursor: pointer;
  fill: $white;
  height: 11px;
  padding: 3px;
  position: absolute;
  right: 12px;
  top: 12px;
  width: 10px;
}
