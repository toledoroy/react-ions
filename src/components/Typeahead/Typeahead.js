import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames/bind'
import enhanceWithClickOutside from 'react-click-outside'
import fuzzy from 'fuzzy'
import debounce from 'lodash/debounce'
import Loader from 'react-loader'
import Input from '../Input'
import Icon from '../Icon'
import style from './style.scss'

export class Typeahead extends React.Component {
  constructor(props) {
    super(props)

    this.onChange = typeof this.props.searchCallback === 'function' && props.searchDebounceTime > 0 ? debounce(this.handleChange, props.searchDebounceTime) : this.handleChange
  }

  static defaultProps = {
    disabled: false,
    options: [],
    valueProp: '',
    displayProp: '',
    resetAfterSelection: false,
    searchDebounceTime: 0
  }

  static propTypes = {
    /**
     * Name of the typeahead.
     */
    name: PropTypes.string,
    /**
     * A string to display as the placeholder text.
     */
    placeholder: PropTypes.string,
    /**
     * An array of objects which will be used as the options for the select field.
     */
    options: PropTypes.array.isRequired,
    /**
     * Value of the typeahead.
     */
    value: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string
    ]),
    /**
     * Which field in the option object will be used as the value of the select field.
     */
    valueProp: PropTypes.string.isRequired,
    /**
     * Which field in the option object will be used as the display of the select field.
     */
    displayProp: PropTypes.string.isRequired,
    /**
     * Whether the select field is disabled.
     */
    disabled: PropTypes.bool,
    /**
     * Optional styles to add to the select field.
     */
    optClass: PropTypes.string,
    /**
     * A callback function to be called when an option is selected.
     */
    changeCallback: PropTypes.func,
    /**
     * A callback for updating options when typeahead search value is changed.
     */
    searchCallback: PropTypes.func,
    /**
     * A loading state to be set to true when asynchronous searching is in progress.
     */
    loading: PropTypes.bool,
    /**
     * A function to filter options.
     */
    optionsFilterPredicate: PropTypes.func,
    /**
     * Clear search string after selection.
     */
    resetAfterSelection: PropTypes.bool,
    /**
     * Search debounce time.
     */
    searchDebounceTime: PropTypes.number,
    /**
     * Text shown above the typeahead.
     */
    label: PropTypes.string,
    /**
     * When set to true, the component (input) will accept a custom value
     */
    allowCustomValue: PropTypes.bool,
    /**
     * A helper will render inline style='width: <value>'.
     */
    width: PropTypes.string
  }

  state = {
    isActive: false,
    value: this.props.value || '',
    results: [],
    selected: '',
    searchStr: this.props.value || ''
  }

  componentWillMount = () => {
    if (typeof this.state.value !== 'undefined' && this.state.value !== '' && this.getIndex(this.state.value, this.props.options) > -1) {
      this.selectItem(this.state.value, this.props.options)
    } else {
      this.setState({selected: ''})
    }
  }

  componentWillReceiveProps = nextProps => {
    const { allowCustomValue, changeCallback } = this.props
    const valueIsEmpty = nextProps.value === ''
    const valueChanged = nextProps.value !== this.state.value
    const searchStringIsEmpty = this.state.searchStr !== ''
    const optionExists = this.getIndex(nextProps.value, nextProps.options) > -1

    // If the option exists select it
    if (nextProps.value && valueChanged && optionExists) {
      this.setState({ value: nextProps.value }, () => {
        this.selectItem(nextProps.value, nextProps.options)
      })
    }
    // Else if allowCustomValue is true trigger the change callback
    else if (nextProps.value && valueChanged && allowCustomValue) {
      this.setState({ value: nextProps.value, searchStr: nextProps.value }, () => {
        changeCallback && changeCallback({ target: { name: nextProps.name, value: nextProps.value } })
      })
    }
    // When the value is an empty string and the current state value is not an empty string
    // Or when the value is an empty string and the search string exists
    // This ensures that 'custom' values are cleared
    else if ((valueIsEmpty && valueChanged) || (allowCustomValue && valueIsEmpty && searchStringIsEmpty)) {
      this.clearSearch()
    }
  }

  selectOption = option => {
    let normalizedOption = option.original ? option.original : option

    let newState = {
      selected: normalizedOption,
      searchStr: normalizedOption[this.props.displayProp],
      value: normalizedOption[this.props.valueProp],
      isActive: false
    }

    if (this.props.resetAfterSelection) {
      newState.searchStr = ''
      newState.value = ''

      // Focus the input field
      this._inputField.focus()
    }

    this.setState(newState, () => {
      if (typeof this.props.changeCallback === 'function') {
        this.props.changeCallback({
          target: {
            name: this.props.name,
            value: normalizedOption[this.props.valueProp],
            option: normalizedOption
          }
        })
      }
    })
  }

  selectItem = (value, options) => {
    let index = this.getIndex(value, options)

    if (index >= 0) {
      this.selectOption(options[index], false)
    }
  }

  getIndex = (value, options) => {
    let optionIndex = -1

    options.map((option, index) => {
      if (option[this.props.valueProp] === value) {
        optionIndex = index
      }
    })
    return optionIndex
  }

  handleChange = event => {
    if (!event.target.value.length) {
      this.clearSearch()
      return
    }

    this.setState({searchStr: event.target.value})

    if (this.props.allowCustomValue) {
      this.props.changeCallback({
        target: {
          name: this.props.name,
          value: event.target.value
        }
      })
    }

    if (typeof this.props.searchCallback === 'function') {
      this.props.searchCallback(event.target.value).then(options => {
        this.updateResults(event, options)
      })
    } else {
      this.updateResults(event, this.props.options)
    }
  }

  handleClickOutside = () => {
    this.setState({isActive: false})
  }

  updateResults = (event, options) => {
    let str = {
      pre: '<b>',
      post: '</b>',
      extract: el => {
        return el[this.props.displayProp]
      }
    }

    if (this.props.optionsFilterPredicate) {
      options = options.filter(this.props.optionsFilterPredicate)
    }

    let results = fuzzy.filter(event.target.value, options, str)

    this.setState({results: results, isActive: true})
  }

  getDynamicList = str => {
    return {
      __html: str
    }
  }

  clearSearch = () => {
    this.setState({isActive: false, searchStr: '', selected: '', value: ''}, () => {
      if (typeof this.props.changeCallback === 'function') {
        this.props.changeCallback({
          target: {
            name: this.props.name,
            value: '',
            option: ''
          }
        })
      }
    })
  }

  render() {
    const cx = classNames.bind(style)
    const loaderClass = this.props.loading ? 'loading' : null
    const typeaheadClass = cx(style['typeahead-component'], loaderClass, this.props.optClass)

    const spinnerOptions = {
      color: '#9198A0',
      length: 4,
      lines: 10,
      radius: 5,
      left: 'calc(100% - 21px)',
      width: 3
    }

    const { placeholder, disabled, loading, label, width } = this.props

    const listWidth = width ? { width: `calc(${width} - 4px)` } : null

    const options = this.state.results.map((option, index) =>
      <li
        key={index}
        onClick={this.selectOption.bind(null, option, true)}
        dangerouslySetInnerHTML={this.getDynamicList(option.string)} />
    )

    return (
      <div className={typeaheadClass}>
        { label && <label>{label}</label> }

        <div className={style['input-wrapper']}>
          <Input 
            ref={c => this._inputField = c} 
            changeCallback={this.onChange} 
            value={this.state.searchStr} 
            placeholder={placeholder} 
            disabled={disabled} 
            width={width}
          />

          { this.state.searchStr !== '' && !loading && !disabled
            ? <Icon name='md-close' onClick={this.clearSearch} className={style['reset-button']}>Reset</Icon>
            : null
          }
        </div>

        { loading ? <Loader loaded={false} options={spinnerOptions} /> : null }

        { this.state.isActive
          ? <ul className={style['typeahead-list']} style={listWidth}>
            {options}
          </ul>
          : null
        }
      </div>
    )
  }
}

export default enhanceWithClickOutside(Typeahead)
