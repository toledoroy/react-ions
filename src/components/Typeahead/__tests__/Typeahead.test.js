import React from 'react'
import { Typeahead } from '../Typeahead'
import sinon from 'sinon'

let wrapper, inst

const options = [
  {value: 'AT', display: 'Austria'},
  {value: 'BM', display: 'Bermuda'},
  {value: 10, display: 'Number'}
]

it('should shallow render itself', () => {
  wrapper = shallow(<Typeahead options={options} valueProp='value' displayProp='display' />)

  expect(wrapper).toMatchSnapshot()
})

it('should have placeholder text', () => {
  wrapper = shallow(<Typeahead options={options} valueProp='value' displayProp='display' placeholder='test' />)

  expect(wrapper).toMatchSnapshot()
})

it('should have a label', () => {
  wrapper = shallow(<Typeahead options={options} valueProp='value' displayProp='display' label='Typeahead Label' />)

  expect(wrapper).toMatchSnapshot()
})

it('should be disabled', () => {
  wrapper = shallow(<Typeahead options={options} valueProp='value' displayProp='display' disabled={true} />)

  expect(wrapper).toMatchSnapshot()
})

it('should have an extra class', () => {
  wrapper = shallow(<Typeahead options={options} valueProp='value' displayProp='display' optClass='test-class' />)

  expect(wrapper).toMatchSnapshot()
})

it('should display a list when the user types a letter', () => {
  wrapper = shallow(<Typeahead options={options} valueProp='value' displayProp='display' />)
  inst = wrapper.instance()

  inst.handleChange({ target: { value: 'a' }})

  wrapper.update()

  expect(wrapper).toMatchSnapshot()
})

it('should display a loader', () => {
  wrapper = shallow(<Typeahead options={options} valueProp='value' displayProp='display' loading={true} />)

  expect(wrapper).toMatchSnapshot()
})

it('should take a search callback', () => {
  const promiseOptions = [
    {value: 'US', display: 'United States'}
  ]
  const searchStub = sinon.stub().returns(Promise.resolve(promiseOptions))

  wrapper = mount(<Typeahead options={options} valueProp='value' displayProp='display' searchCallback={searchStub} />)
  wrapper.find('input').simulate('change', {target: {value: 'b'}})

  expect(searchStub.calledWithExactly('b')).toBe(true)
})

it('should have a custom search debounce time', done => {
  const promiseOptions = [
    {value: 'US', display: 'United States'}
  ]
  const searchStub = sinon.stub().returns(Promise.resolve(promiseOptions))

  wrapper = mount(<Typeahead options={options} valueProp='value' displayProp='display' searchCallback={searchStub} searchDebounceTime={200} />)
  wrapper.find('input').simulate('change', {target: {value: 'b'}})

  expect(searchStub.calledWithExactly('b')).toBe(false)

  setTimeout(() => {
    expect(searchStub.calledWithExactly('b')).toBe(true)
    done()
  }, 400)
})

it('should update when props are set', () => {
  wrapper = shallow(<Typeahead options={options} valueProp='value' displayProp='display' value={10} optClass='test-class' />)

  expect(wrapper).toMatchSnapshot()
})

it('should not clear search string after selection', () => {
  wrapper = shallow(<Typeahead options={options} valueProp='value' displayProp='display' />)

  inst = wrapper.instance()

  wrapper.instance().handleChange({ target: { value: 'a' } })

  expect(wrapper.state('searchStr')).toEqual('a')

  wrapper.instance().selectOption(options[0])
  expect(wrapper.state('searchStr')).toEqual('Austria')
})

it('should clear search string after selection', () => {
  wrapper = mount(<Typeahead resetAfterSelection={true} options={options} valueProp='value' displayProp='display' />)

  inst = wrapper.instance()

  inst.handleChange({ target: { value: 'a' } })

  expect(wrapper.state('searchStr')).toEqual('a')

  inst.selectOption(options[0])
  expect(wrapper.state('searchStr')).toEqual('')
})

it('should clear search when the clear button is clicked', () => {
  wrapper = mount(<Typeahead resetAfterSelection={true} options={options} valueProp='value' displayProp='display' changeCallback={sinon.spy()} />)

  inst = wrapper.instance()

  inst.handleChange({ target: { value: 'a' } })

  expect(wrapper.state('searchStr')).toEqual('a')

  inst.clearSearch()
  expect(wrapper.state('searchStr')).toEqual('')
})

it('should clear search string when input value is an empty string', () => {
  const changeCallback = sinon.spy()

  wrapper = mount(<Typeahead name='typeahead' resetAfterSelection={true} options={options} valueProp='value' displayProp='display' changeCallback={changeCallback} />)

  let inputField = wrapper.find('input')

  inputField.simulate('change', {target: {value: 'a'}})
  expect(wrapper.state().searchStr).toEqual('a')

  inputField.simulate('change', {target: {value: ''}})
  expect(wrapper.state().searchStr).toEqual('')
  expect(changeCallback.calledWithExactly({ target: { name: 'typeahead', value: '', option: '' } })).toBe(true)
})

it('should set state when props are received', () => {
  const changeCallback = sinon.spy()

  wrapper = mount(<Typeahead name='typeahead' options={options} valueProp='value' displayProp='display' changeCallback={changeCallback} />)

  wrapper.setProps({ value: 'AT' })

  expect(wrapper.state().selected.value).toEqual('AT')
  expect(wrapper.state().selected.display).toEqual('Austria')

  wrapper.setProps({ value: '' })

  expect(wrapper.state().selected).toEqual('')
})

it('should allow a custom value', () => {
  const changeCallback = sinon.spy()

  wrapper = mount(<Typeahead name='typeahead' options={options} valueProp='value' displayProp='display' allowCustomValue={true} changeCallback={changeCallback} />)

  let inputField = wrapper.find('input')

  inputField.simulate('change', {target: {value: 'a random string'}})
  expect(changeCallback.calledWithExactly({ target: { name: 'typeahead', value: 'a random string' } })).toBe(true)
})

it('should set state when the value changes and a custom value is allowed', () => {
  const changeCallback = sinon.spy()

  wrapper = shallow(<Typeahead name='typeahead' options={options} valueProp='value' displayProp='display' allowCustomValue={true} changeCallback={changeCallback} />)
  const inst = wrapper.instance()

  inst.componentWillReceiveProps({ value: 'a random string', options })
  expect(wrapper.state().value).toEqual('a random string')
})