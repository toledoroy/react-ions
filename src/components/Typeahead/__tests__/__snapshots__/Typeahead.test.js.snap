// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be disabled 1`] = `
<div
  className="typeahead-component"
>
  <styled.div
    className="typeahead-component-styled"
    css={
      Object {
        "position": "relative",
        "width": undefined,
      }
    }
  >
    <div
      className="input-wrapper"
    >
      <Input
        changeCallback={[Function]}
        disabled={true}
        type="text"
        value=""
        valueType="string"
      />
    </div>
    <Spinner
      className="typeahead-spinner"
      size="16"
      type="spinner-circular"
    />
  </styled.div>
</div>
`;

exports[`should display a list when the user types a letter 1`] = `
<div
  className="typeahead-component"
>
  <styled.div
    className="typeahead-component-styled"
    css={
      Object {
        "position": "relative",
        "width": undefined,
      }
    }
  >
    <div
      className="input-wrapper"
    >
      <Input
        changeCallback={[Function]}
        disabled={false}
        type="text"
        value="a"
        valueType="string"
      />
      <Icon
        className="reset-button"
        height="24"
        name="md-close"
        onClick={[Function]}
        width="24"
      >
        Reset
      </Icon>
    </div>
    <Spinner
      className="typeahead-spinner"
      size="16"
      type="spinner-circular"
    />
    <ul
      className="typeahead-list"
      style={null}
    >
      <li
        dangerouslySetInnerHTML={
          Object {
            "__html": "<b>A</b>ustria",
          }
        }
        key="0"
        onClick={[Function]}
      />
      <li
        dangerouslySetInnerHTML={
          Object {
            "__html": "Bermud<b>a</b>",
          }
        }
        key="1"
        onClick={[Function]}
      />
    </ul>
  </styled.div>
</div>
`;

exports[`should display a loader 1`] = `
<div
  className="typeahead-component loading"
>
  <styled.div
    className="typeahead-component-styled"
    css={
      Object {
        "position": "relative",
        "width": undefined,
      }
    }
  >
    <div
      className="input-wrapper"
    >
      <Input
        changeCallback={[Function]}
        disabled={false}
        type="text"
        value=""
        valueType="string"
      />
    </div>
    <Spinner
      className="typeahead-spinner"
      loading={true}
      size="16"
      type="spinner-circular"
    />
  </styled.div>
</div>
`;

exports[`should have a label 1`] = `
<div
  className="typeahead-component"
>
  <label>
    Typeahead Label
  </label>
  <styled.div
    className="typeahead-component-styled"
    css={
      Object {
        "position": "relative",
        "width": undefined,
      }
    }
  >
    <div
      className="input-wrapper"
    >
      <Input
        changeCallback={[Function]}
        disabled={false}
        type="text"
        value=""
        valueType="string"
      />
    </div>
    <Spinner
      className="typeahead-spinner"
      size="16"
      type="spinner-circular"
    />
  </styled.div>
</div>
`;

exports[`should have an extra class 1`] = `
<div
  className="typeahead-component test-class"
>
  <styled.div
    className="typeahead-component-styled"
    css={
      Object {
        "position": "relative",
        "width": undefined,
      }
    }
  >
    <div
      className="input-wrapper"
    >
      <Input
        changeCallback={[Function]}
        disabled={false}
        type="text"
        value=""
        valueType="string"
      />
    </div>
    <Spinner
      className="typeahead-spinner"
      size="16"
      type="spinner-circular"
    />
  </styled.div>
</div>
`;

exports[`should have placeholder text 1`] = `
<div
  className="typeahead-component"
>
  <styled.div
    className="typeahead-component-styled"
    css={
      Object {
        "position": "relative",
        "width": undefined,
      }
    }
  >
    <div
      className="input-wrapper"
    >
      <Input
        changeCallback={[Function]}
        disabled={false}
        placeholder="test"
        type="text"
        value=""
        valueType="string"
      />
    </div>
    <Spinner
      className="typeahead-spinner"
      size="16"
      type="spinner-circular"
    />
  </styled.div>
</div>
`;

exports[`should shallow render itself 1`] = `
<div
  className="typeahead-component"
>
  <styled.div
    className="typeahead-component-styled"
    css={
      Object {
        "position": "relative",
        "width": undefined,
      }
    }
  >
    <div
      className="input-wrapper"
    >
      <Input
        changeCallback={[Function]}
        disabled={false}
        type="text"
        value=""
        valueType="string"
      />
    </div>
    <Spinner
      className="typeahead-spinner"
      size="16"
      type="spinner-circular"
    />
  </styled.div>
</div>
`;

exports[`should update when props are set 1`] = `
<div
  className="typeahead-component test-class"
>
  <styled.div
    className="typeahead-component-styled"
    css={
      Object {
        "position": "relative",
        "width": undefined,
      }
    }
  >
    <div
      className="input-wrapper"
    >
      <Input
        changeCallback={[Function]}
        disabled={false}
        type="text"
        value="Number"
        valueType="string"
      />
      <Icon
        className="reset-button"
        height="24"
        name="md-close"
        onClick={[Function]}
        width="24"
      >
        Reset
      </Icon>
    </div>
    <Spinner
      className="typeahead-spinner"
      size="16"
      type="spinner-circular"
    />
  </styled.div>
</div>
`;
